scapy/contrib/tzsp.py

```diff
@@ -0,0 +1,493 @@
+        ByteEnumField('type', _TZSPTag.TAG_TYPE_RAW_RSSI, _TZSPTag.TAG_TYPES),
+        ByteField('len', 1),
+        ByteField('raw_rssi', 0)
+    ]
+
+
+class TZSPTagRawRSSIShort(_TZSPTag):
+    """
+    relative received signal strength - signed short value
+    """
+    fields_desc = [
+        ByteEnumField('type', _TZSPTag.TAG_TYPE_RAW_RSSI, _TZSPTag.TAG_TYPES),
+        ByteField('len', 2),
+        ShortField('raw_rssi', 0)
+    ]
+
+
+class TZSPTagSNRByte(_TZSPTag):
+    """
+    signal noise ratio - signed byte value
+    """
+    fields_desc = [
+        ByteEnumField('type', _TZSPTag.TAG_TYPE_SNR, _TZSPTag.TAG_TYPES),
+        ByteField('len', 1),
+        ByteField('snr', 0)
+    ]
+
+
+class TZSPTagSNRShort(_TZSPTag):
+    """
+    signal noise ratio - signed short value
+    """
+    fields_desc = [
+        ByteEnumField('type', _TZSPTag.TAG_TYPE_SNR, _TZSPTag.TAG_TYPES),
+        ByteField('len', 2),
+        ShortField('snr', 0)
+    ]
+
+
+class TZSPTagDataRate(_TZSPTag):
+    """
+    wireless link data rate
+    """
+    DATA_RATE_UNKNOWN = 0x00
+    DATA_RATE_1 = 0x02
+    DATA_RATE_2 = 0x04
+    DATA_RATE_5_5 = 0x0B
+    DATA_RATE_6 = 0x0C
+    DATA_RATE_9 = 0x12
+    DATA_RATE_11 = 0x16
+    DATA_RATE_12 = 0x18
+    DATA_RATE_18 = 0x24
+    DATA_RATE_22 = 0x2C
+    DATA_RATE_24 = 0x30
+    DATA_RATE_33 = 0x42
+    DATA_RATE_36 = 0x48
+    DATA_RATE_48 = 0x60
+    DATA_RATE_54 = 0x6C
+    DATA_RATE_LEGACY_1 = 0x0A
+    DATA_RATE_LEGACY_2 = 0x14
+    DATA_RATE_LEGACY_5_5 = 0x37
+    DATA_RATE_LEGACY_11 = 0x6E
+
+    DATA_RATES = {
+        DATA_RATE_UNKNOWN: 'unknown',
+        DATA_RATE_1: '1 MB/s',
+        DATA_RATE_2: '2 MB/s',
+        DATA_RATE_5_5: '5.5 MB/s',
+        DATA_RATE_6: '6 MB/s',
+        DATA_RATE_9: '9 MB/s',
+        DATA_RATE_11: '11 MB/s',
+        DATA_RATE_12: '12 MB/s',
+        DATA_RATE_18: '18 MB/s',
+        DATA_RATE_22: '22 MB/s',
+        DATA_RATE_24: '24 MB/s',
+        DATA_RATE_33: '33 MB/s',
+        DATA_RATE_36: '36 MB/s',
+        DATA_RATE_48: '48 MB/s',
+        DATA_RATE_54: '54 MB/s',
+        DATA_RATE_LEGACY_1: '1 MB/s (legacy)',
+        DATA_RATE_LEGACY_2: '2 MB/s (legacy)',
+        DATA_RATE_LEGACY_5_5: '5.5 MB/s (legacy)',
+        DATA_RATE_LEGACY_11: '11 MB/s (legacy)',
+    }
+
+    fields_desc = [
+        ByteEnumField('type', _TZSPTag.TAG_TYPE_DATA_RATE, _TZSPTag.TAG_TYPES),
+        ByteField('len', 1),
+        ByteEnumField('data_rate', DATA_RATE_UNKNOWN, DATA_RATES)
+    ]
+
+
+class TZSPTagTimestamp(_TZSPTag):
+    """
+    MAC receive timestamp
+    """
+    fields_desc = [
+        ByteEnumField('type', _TZSPTag.TAG_TYPE_TIMESTAMP, _TZSPTag.TAG_TYPES),
+        ByteField('len', 4),
+        IntField('timestamp', 0)
+    ]
+
+
+class TZSPTagContentionFree(_TZSPTag):
+    """
+    packet received in contention free period
+    """
+    NO = 0x00
+    YES = 0x01
+
+    CONTENTION_STATES = {key: ('yes' if key != 0x00 else 'no') for key in range(0x00, 0x100)}
```

Comment:
This is extremely ugly. As you use the same thing several times, I suggest you to use something such as:
```
class YesNoByteField(ByteField):
    __slots__ = [“yes”]
    def __init__(self, name, default, yes=0, *args, **kargs):
        self.yes = yes
        ByteField.__init__(self, name, default, *args, **kargs)
    def i2repr(self, pkt, x):
        return “yes” if x == self.yes else “no”
```

Which you may want to add to `fields.py` for further usages, as quite common

Reply:
good idea. as TZSP also needs a tri-state-representation (0 - no, 1 - yes, 2..ff - reserved) I added a slightly more flexible version than your proposed solution (ed19f9b).

Reply:
Good job !